content/5_Plat/2DRQ.mdx

```diff
@@ -322,93 +348,87 @@ And you might use it like so:
 using ll = long long;
 
-// BeginCodeSnip{BIT2D}
-// index of largest value <= x in v (sorted)
-// if v = [1, 2, 4], ind(v, 3) would return 1
-int ind(vector<int> v, int x) {
-	return upper_bound(v.begin(), v.end(), x) - v.begin() - 1;
-}
-
-class BIT2D {
+// BeginCodeSnip{Offline 2D BIT (from the module)}
+template <typename T> class OfflineBIT2D {
   private:
-	int n;  // max x-coordinate
-	vector<vector<int>> vals, bit;
+	const int n;
+	vector<vector<int>> vals;
+	vector<vector<T>> bit;
+
+	/** @return the first index i such that v[i] <= x */
+	int ind(const vector<int> &v, int x) {
+		return upper_bound(begin(v), end(v), x) - begin(v) - 1;
+	}
 
   public:
-	BIT2D(int n, vector<pair<int, int>> &todo) : n(n), vals(n + 1), bit(n + 1) {
-		// sort points by y-coordinate
-		sort(todo.begin(), todo.end(),
-		     [](pair<int, int> a, pair<int, int> b) { return a.second < b.second; });
-		// ensures vals and bit are 1-indexed
+	OfflineBIT2D(int n, vector<array<int, 2>> &todo) : n(n), vals(n + 1), bit(n + 1) {
+		sort(begin(todo), end(todo),
+		     [](const array<int, 2> &a, const array<int, 2> &b) -> bool {
+			     return a[1] < b[1];
+		     });
+
 		for (int i = 1; i <= n; i++) { vals[i].push_back(0); }
-		for (auto [x, y] : todo) {
-			for (int z = x; z <= n; z += z & -z) {
-				if (vals[z].back() != y) { vals[z].push_back(y); }
+		for (auto [r, c] : todo) {
+			r++, c++;
+			for (; r <= n; r += r & -r) {
+				if (vals[r].back() != c) { vals[r].push_back(c); }
 			}
 		}
 		for (int i = 1; i <= n; i++) { bit[i].resize(vals[i].size()); }
 	}
 
-	/** adds t to the point (x, y) */
-	void upd(int x, int y, int t = 1) {
-		for (; x <= n; x += x & -x) {
-			int z = ind(vals[x], y);
-			assert(z && vals[x][z] == y);
-			for (; z < bit[x].size(); z += z & -z) { bit[x][z] += t; }
+	/** adds val to the point (r, c) */
+	void add(int r, int c, T val) {
+		r++, c++;
+		for (; r <= n; r += r & -r) {
+			int i = ind(vals[r], c);
+			for (; i < bit[r].size(); i += i & -i) { bit[r][i] += val; }
 		}
 	}
 
-	/** @return sum of points in rectangle with top-right corner (x, y) */
-	int query(int x, int y) {
-		int tot = 0;
-		for (; x > 0; x -= x & -x) {
-			for (int z = ind(vals[x], y); z > 0; z -= z & -z) { tot += bit[x][z]; }
+	/** @returns sum of points with row in [0, r] and column in [0, c] */
+	T rect_sum(int r, int c) {
+		r++, c++;
+		T sum = 0;
+		for (; r > 0; r -= r & -r) {
+			int i = ind(vals[r], c);
+			for (; i > 0; i -= i & -i) { sum += bit[r][i]; }
 		}
-		return tot;
+		return sum;
 	}
 
-	/** @returns sum of points with x in [x1, x2] and y in [y1, y2] */
-	int query(int x1, int x2, int y1, int y2) {
-		if (x1 > x2 || y1 > y2) { return 0; }
-		int tr = query(x2, y2);          // top-right
-		int tl = query(x1 - 1, y2);      // top-left
-		int br = query(x2, y1 - 1);      // bottom-right
-		int bl = query(x1 - 1, y1 - 1);  // bottom-left
-		return tr - tl - br + bl;
+	/** @returns sum of points with r in [r1, r2] and c in [c1, c2] */
```

Comment:
no need for doc comments since they can go to the previous imp

Reply:
ic